sdk/communication/Azure.Communication.Common/src/CommunicationIdentifier.cs

```diff
@@ -3,20 +3,61 @@
 
 using System;
 using System.ComponentModel;
+using Azure.Core;
 
 namespace Azure.Communication
 {
     /// <summary>Represents an identifier in Azure Communication Services.</summary>
-#pragma warning disable CS0659 // Type overrides Object.Equals(object o) but does not override Object.GetHashCode()
     public abstract class CommunicationIdentifier : IEquatable<CommunicationIdentifier>
-#pragma warning restore CS0659 // Type overrides Object.Equals(object o) but does not override Object.GetHashCode()
     {
+        /// <summary>
+        /// Returns the RawId representation of the CommunicationIdentifier.
+        /// You can use the RawId for encoding the identifier and then use it as a key in a database.
+        /// </summary>
+        public virtual string RawId { get; }
```

Comment:
@DominikMe `abstract` would be a breaking change. you ok with this?

Reply:
Technically it would be breaking because we hadn't marked CommunicationIdentfier as `sealed`. But I don't expect that any customer has sub-classed CommunicationIdentifier as we wouldn't know how to treat this sub-class. Maybe for testing, but customers would have probably used a mocking technique.
`abstract` is more correct, as we need each identifier to define how they encode themselves, so that is my preference.

@christothes or @tg-msft can you advise whether we can make an exception here and use `abstract`?

Reply:
No, either changing the existing modifier or adding a new abstract member we forwarded this to would be an unacceptable breaking change per the Framework Design Guidelines (see D.11.21 and D.7.2).  I think we'll have to live with the design and just document the expectations.

Reply:
:-( That's unfortunate, I cannot imagine that we would break anybody in reality. This class was never designed to be sub-classed by others.

If we have to live with it for strict reasons, we should maybe throw in `CommunicationIdentifier`'s implementation to discover bugs early where a sub-class doesn't override `RawId` as it's supposed to.

Reply:
>we should maybe throw in CommunicationIdentifier's implementation to discover bugs early where a sub-class doesn't override RawId as it's supposed to.

yeah, but if we base `GetHashCode()` on it, then it's effectively the same as creating a new abstract member (the derived class is forced to implement it). Is the fact that the class was never designed to be sub-classed by others a good enough justification to do so? @tg-msft @DominikMe

Reply:
yes, we can definitely bring up the topic on Thursday! thanks for your comments so far 🙂 👍

Reply:
@DominikMe 
We discussed it on Thursday (recording [here](https://microsoft-my.sharepoint.com/:v:/p/teglaza/EUmWLrxWxQVEqUsJs4_eceYB7YQ1HqCMPeQsqf9xOA2pxA) - it was the first topic of the session) and there were two suggestions for how to tackle this issue, both of which assume adding default implementation:

1. return a constant `UnknownIdentifier` which would equal to an identifier that would not collide with any other identifier (like `0:unknown`)
2. return `null`


The first solution is better in the case we want to always have some value available that we can insert into the database. If not, we should go with the second approach - returning null. Since we're doing this also to support inserting the IDs to a DB, I'm slightly in favor of the first way. How about you @DominikMe ?

As for making sure that our subclasses implement `RawId`, I suggest adding an automatic test that would check that via reflection.

Also, we need to update the docs, to instruct Contoso to implement this member if they have a derived class.

Reply:
To me, nobody should ever use the return value of the base class. If a sub-class doesn't implement this "virtual" method it's a bug. Having a string like `0:unknown` imo just hides the error. If this string gets sent to our services, it will just fail there. If the string gets persisted in the customer's db it won't be better than `null` because it's meaningless.

> Also, we need to update the docs, to instruct Contoso to implement this member if they have a derived class.
`CommunicationIdentifier` really isn't meant to be subclassed by Contoso and we shouldn't encourage them. The identifier is a 1:1 mapping to values that are entirely defined in the IC3 services.

Reply:
keeping null + adding a test for subclasses implementing the member